crates/builder/src/graph/adj_list.rs

```diff
@@ -59,12 +92,119 @@ impl<NI: Idx> AdjacencyList<NI, ()> {
             std::mem::align_of::<Target<NI, ()>>(),
             std::mem::align_of::<NI>()
         );
-
-        let targets = self.edges[node.index()].as_slice();
-
         // SAFETY: The types Target<T, ()> and T are verified to have the same
         //         size and alignment.
-        unsafe { std::slice::from_raw_parts(targets.as_ptr().cast(), targets.len()) }
+        unsafe { std::slice::from_raw_parts(self.targets.as_ptr().cast(), self.targets.len()) }
+    }
+}
+
+pub struct TargetsIter<'slice, NI: Idx> {
+    _targets: Targets<'slice, NI>,
+    slice: std::slice::Iter<'slice, NI>,
+}
+
+impl<'slice, NI: Idx> TargetsIter<'slice, NI> {
+    pub fn as_slice(&self) -> &'slice [NI] {
+        self.slice.as_slice()
+    }
+}
+
+impl<'slice, NI: Idx> IntoIterator for Targets<'slice, NI> {
+    type Item = &'slice NI;
+
+    type IntoIter = TargetsIter<'slice, NI>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        let slice = self.as_slice();
+        TargetsIter {
+            _targets: self,
+            slice: slice.iter(),
+        }
+    }
+}
+
+impl<'slice, NI: Idx> Iterator for TargetsIter<'slice, NI> {
+    type Item = &'slice NI;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.slice.next()
+    }
+}
+
+impl<NI: Idx> AdjacencyList<NI, ()> {
+    #[inline]
+    pub(crate) fn targets<'list, 'slice>(&'list self, node: NI) -> Targets<'slice, NI>
+    where
+        'list: 'slice,
+    {
+        let targets = self.edges[node.index()]
+            .lock()
+            .expect("Could not acquire lock");
+
+        Targets { targets }
+    }
+}
+
+#[derive(Debug)]
+pub struct TargetsWithValues<'slice, NI: Idx, EV> {
+    targets: MutexGuard<'slice, Vec<Target<NI, EV>>>,
+}
+
+impl<'slice, NI: Idx, EV> TargetsWithValues<'slice, NI, EV> {
+    pub fn as_slice(&self) -> &'slice [Target<NI, EV>] {
+        // SAFETY: We can upcast the lifetime since the MutexGuard
+        // is not exposed, so it is not possible to deref mutable.
```

Comment:
This comment can also be added to the other SAFETY note above for the no edge values Targets impl